src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts

```diff
@@ -0,0 +1,363 @@
+                    this._msalApplication.loginRedirect(request);
+                } catch (e) {
+                    return e;
+                }
+            } else {
+                return e;
+            }
+        }
+    }
+
+    completeSignIn() {
+        return this._callbackPromise;
+    }
+
+    async signOut(state: any) {
+        // We are about to sign out, so clear any state before we do so and leave just the sign out state for
+        // the current sign out flow.
+        this.purgeState();
+
+        const logoutStateId = await this.saveState(state);
+
+        // msal.js doesn't support providing logout state, so we shim it by putting the identifier in session storage
+        // and using that on the logout callback to workout the problems.
+        sessionStorage.setItem(`${AuthenticationService._infrastructureKey}.LogoutState`, logoutStateId);
+
+        this._msalApplication.logout();
+
+        // We are about to be redirected.
+        return this.redirect();
+    }
+
+    async completeSignOut(url: string) {
+        const logoutStateId = sessionStorage.getItem(`${AuthenticationService._infrastructureKey}.LogoutState`);
+        const updatedUrl = new URL(url);
+        updatedUrl.search = `?state=${logoutStateId}`;
+        const logoutState = await this.retrieveState(updatedUrl.href, /*isLogout*/ true);
+
+        sessionStorage.removeItem(`${AuthenticationService._infrastructureKey}.LogoutState`);
+
+        if (logoutState) {
+            return this.success(logoutState);
+        } else {
+            return this.operationCompleted();
+        }
+    }
+
+    // msal.js only allows a string as the account state and it simply attaches it to the sign-in request state.
+    // Given that we don't want to serialize the entire state and put it in the query string, we need to serialize the
+    // state ourselves and pass an identifier to retrieve it while in the callback flow.
+    async saveState<T>(state: T): Promise<string> {
+        const base64UrlIdentifier = await new Promise<string>((resolve, reject) => {
+            const reader = new FileReader();
+            reader.onloadend = evt => resolve((evt?.target?.result as string)
+                // The result comes back as a base64 string inside a dataUrl.
+                // We remove the prefix and convert it to base64url by replacing '+' with '-', '/' with '_' and removing '='.
+                .split(',')[1].replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, ''));
+            reader.onerror = evt => reject(evt.target?.error?.message);
+
+            // We generate a base 64 url encoded string of random data.
+            const entropy = window.crypto.getRandomValues(new Uint8Array(32));
+            reader.readAsDataURL(new Blob([entropy]));
+        });
+
+        sessionStorage.setItem(`${AuthenticationService._infrastructureKey}.AuthorizeService.${base64UrlIdentifier}`, JSON.stringify(state));
+        return base64UrlIdentifier;
```

Comment:
What's the reason for using `FileReader` and representing the identifier as a base64url? If there was some other way to generate a sufficiently unique random string, would that be a sufficient alternative, or does it have to be in this kind of format for some reason?

For example, what if you used `crypto.getRandomValues(new Uint8Array(32))` and then concatenated them into a string of hex values, all synchronously?

Reply:
Didn't think of using hex to be honest. The base64 version is a bit shorter, and I just happened to know how to do this.

I'm going to see if the msal folks can manage the state within the library themselves, so hopefully this logic goes away.